kernel/src/snapshot.rs

```diff
@@ -471,6 +506,62 @@ mod tests {
         assert!(cp.is_none())
     }
 
+    #[test]
+    fn test_read_log_with_out_of_date_last_checkpoint() {
+        let store = Arc::new(InMemory::new());
+
+        fn get_path(index: usize, suffix: &str) -> Path {
+            let path = format!("_delta_log/{index:020}.{suffix}");
+            Path::from(path.as_str())
+        }
+        let data = bytes::Bytes::from("kernel-data");
+
+        // add log files to store
+        tokio::runtime::Runtime::new()
+            .expect("create tokio runtime")
+            .block_on(async {
+                for path in [
+                    get_path(0, "json"),
+                    get_path(1, "checkpoint.parquet"),
+                    get_path(2, "json"),
+                    get_path(3, "checkpoint.parquet"),
+                    get_path(4, "json"),
+                    get_path(5, "checkpoint.parquet"),
+                    get_path(6, "json"),
+                    get_path(7, "json"),
```

Comment:
could add a test with stale `_last_checkpoint` or add one in here? (or just open issue as little follow-up)

Reply:
I mean, the stale last checkpoint will just create an out of date `CheckpointMetadata` which is what we do below. 

We already test `read_last_checkpoint`. And if I _do_ create the out of date `_last_checkpoint` file the only way to trigger reading it is just by creating a `Snapshot` which makes it harder to validate exactly what happens.

We could have a follow-up to have such a table on disk, but I'm not sure it has much benefit and I want to avoid having too many of those. :)

I did add in a `_last_checkpoint` file to ensure it doesn't mess anything up.

Reply:
sgtm!!